design/20240918.delayed-certificate-activation.md

```diff
@@ -0,0 +1,140 @@
+This proposal introduces a mechanism to specify a future `NotBefore` time in cert-manager certificates, enabling delayed activation of issued certificates.
+
+## Motivation
+
+Currently, cert-manager creates the Certificate Signing Request (CSR) with `NotBefore` set to the current time (`time.Now()`).  This prevents use cases where a certificate needs to be issued in advance but only become valid at a specific future time. For example,
+
+- **Pre-provisioning Certificates:** Generating certificates for resources that will be deployed in the future, such as for a scheduled application rollout.
+- **Compliance Requirements:** Adhering to specific security policies that mandate a future validation start date for certificates.
+
+Supporting a future `NotBefore` time aligns with the capabilities of x509.CertificateRequest and provides greater flexibility in certificate management.
+
+### Goals
+
+- Extend the `Certificate` and `CertificateRequest` APIs to allow specifying a future `NotBefore` time.
+- Update the `CertificateRequest` controller to honor the specified NotBefore time when creating x509.CertificateRequests.
+- Maintain backward compatibility with existing certificate issuance workflows.
+
+### Non-Goals
+
+- This proposal does not aim to change the behavior of existing certificates without a `NotBefore` field.
+
+## Proposal
+
+Introduce a new optional field, `NotBefore`, to the spec of both `Certificate` and `CertificateRequest` resources. This field will allow users to explicitly define the start time of the certificate's validity period.
+
+### API changes
+
+```
+type CertificateSpec struct {
+  // EXISTING FIELDS
+  // ...
+
+  // NotBefore defines the time when the certificate becomes valid.
+  // If omitted or set to a time in the past, the certificate will be 
+  // valid immediately upon issuance.
+  // +optional
+  NotBefore *metav1.Time `json:"notBefore,omitempty"`
+}
+
+type CertificateRequestSpec struct {
+  // EXISTING FIELDS
+  // ...
+
+  // NotBefore defines the time when the certificate becomes valid.
+  // If omitted or set to a time in the past, the certificate will be 
+  // valid immediately upon issuance.
+  // +optional
+  NotBefore *metav1.Time `json:"notBefore,omitempty"` 
+}
+```
+
+### Examples
+
+#### Issuance with Future NotBefore
+
+```
+apiVersion: cert-manager.io/v1
+kind: Certificate
+metadata:
+  name: example-cert
+spec:
+  secretName: example-cert-tls
+  issuerRef:
+    name: letsencrypt-prod
+    kind: ClusterIssuer
+  commonName: example.com
+  notBefore: "2024-10-01T10:00:00Z" 
+  duration: 2160h # 90 days
+```
+
+This example creates a certificate that will only become valid on October 1st, 2024, even if it is issued earlier.
+
+#### Renewal with NotBefore
+
+When a certificate with a future `NotBefore` is renewed, cert-manager should preserve the original `NotBefore` time if it's still in the future. If the original `NotBefore` is in the past, the renewed certificate should be valid immediately.
```

Comment:
What should the value of `NotBefore` be if the certificate's `NotBefore` was in the past?

Reply:
certificate's `NotBefore` value will depend on the value of `NotBefore` field set in `Certificate` or `CertificateRequest` CR. If the `NotBefore` in spec is set as nil, certificate's `NotBefore` will be set as `time.Now()`.  If the `NotBefore` in spec is set as future time, certificate's `NotBefore` will be set as `spec.NotBefore`. If the `NotBefore` in spec is set as past time, certificate's `NotBefore` will be set as `time.Now()`.

From the implementation perspective, [this](https://github.com/jaingeet/cert-manager/blob/master/pkg/util/pki/certificatetemplate.go#L56-L62) function will be modified as below  - 
```
func CertificateTemplateOverrideDuration(duration time.Duration, notBefore *metav1.Time) CertificateTemplateValidatorMutator {
	return func(req *x509.CertificateRequest, cert *x509.Certificate) error {
		if notBefore != nil && notBefore.Time.After(time.Now()) {
			cert.NotBefore = notBefore.Time
		} else {
			cert.NotBefore = time.Now()
		}

		cert.NotAfter = cert.NotBefore.Add(duration)
		return nil
	}
}
```

Reply:
That's a reasonable behavior, but not obvious from line 115. Could you adjust:

> If the original `NotBefore` is in the past

to clarify this bit?

Reply:
Note that you really shouldn't call `time.Now()` twice like that. It should be called once:
```go
		now := time.Now()
		if notBefore != nil && notBefore.Time.After(now) {
			cert.NotBefore = notBefore.Time
		} else {
			cert.NotBefore = now
		}
```

Reply:
Updated the doc with the suggestion.

Reply:
I think I'd change `valid immediately` to `set to now`.